src/util.js

```diff
@@ -0,0 +1,193 @@
+    const filterTagItems = filter[filterGroupName];
+    // Add all items except the one to be removed
+    const itemsToAdd = filterTagItems.filter(filterTagName => !(filterGroupName === groupName && filterTagName === tagName));
+    if(itemsToAdd.length === 0) return res;
+    return mergeObjects(res, {[filterGroupName]: itemsToAdd});
+  }, {});
+}
+
+/**
+ * Get courses that has at least one lesson that matches filter
+ * @param {Array} courses
+ * @param {Object} filter
+ * @returns {Array} filtered courses
+ */
+export function filterCourses(courses, filter) {
+  const coursesWithFilteredLessons = courses.map(course => {
+    const newCourse = Object.assign({}, course);
+    newCourse.lessons = filterLessons(course.lessons, filter);
+    return newCourse;
+  });
+
+  // Find and sort courses that have at least one lesson that matches filter
+  return coursesWithFilteredLessons.filter((course) => {
+    return course.lessons.length > 0;
+  });
+
+}
+
+/**
+ * Get lessons that have all tags that exist in filter
+ * @param {Array} lessons
+ * @param {Object} filter
+ * @returns {Array} filtered lessons
+ */
+export function filterLessons(lessons, filter) {
+  // Find lessons that matches filter
+  return lessons.filter((lesson) => {
+    return lessonHasTags(lesson, filter);
+  });
+}
+
+///////////////////////////////////
+//////// HELPER FUNCTIONS /////////
+///////////////////////////////////
+
+/**
+ * Fix invalid tags
+ * @param {Object} tags
+ * @returns {Object} valid tags
+ */
+export function cleanseTags(tags) {
+  if(tags == null) return {};
+
+  return Object.keys(tags).reduce((result, groupName) => {
+    const tagItems = fixNonArrayTagList(tags[groupName]);
+    // Ignore tagGroups with no tagItems
+    if(tagItems.length === 0) return result;
+    result[groupName] = tagItems;
+    return result;
+  }, {});
+}
+
+/**
+ * Make sure tagItems is an array. Try to convert to array if it is not.
+ * This happens if tags is created as string or numbers (e.g. someGroupName: tag1, tag2, 12345)
+ * instead of list (e.g. someGroupName: [tag1, tag2, 12345]) in YAML
+ * @param tagItems
+ * @returns {Array}
+ */
+export function fixNonArrayTagList(tagItems) {
+  if(tagItems == null) return [];
+  if (typeof tagItems === 'string') return tagItems.split(/,\s*/);
+  if (typeof  !Array.isArray(tagItems) && tagItems !== 'string') return fixNonArrayTagList(tagItems.toString());
+  return tagItems;
+}
+
+/**
+ * Check if lesson contains all tags in filterTags
+ * @param {Object} lesson
+ * @param {Object} filterTags
+ * @returns {boolean} lesson contains all tags required by filter
+ */
+export function lessonHasTags(lesson, filterTags) {
+  // Filter is empty
+  if(Object.keys(filterTags).length === 0) return true;
+
+  const lessonTags = lesson.tags;
+  for(let groupName in filterTags){
+    if(!lessonTags.hasOwnProperty(groupName) || !filterTags.hasOwnProperty(groupName))return false;
```

Comment:
Dette tror jeg blir feil.

Se på følgende eksempel (testes f.eks. på jsfiddle.net, åpne inspector og se i console):
```
Object.prototype.hi=function(){ console.log('hi'); }
const buz={ foo: 'bar' };

for (let k in buz) {
	if (buz.hasOwnProperty(k)) {
  	console.log('hasOwn: ' + k);
  } else {
    console.log('has not: ' + k);
  }
}
```
Så problem med koden er at dersom noen har, et annet sted i koden, utvidet Object.prototype med f.eks. Object.prototype.hi, så vil denne funksjonen returnere false om den kommer til 'hi', selv om det ikke er en tag. (Altså, groupName blir 'hi', men hasOwnProperty blir false for både lessonTags og filterTags).

Kanskje det er lettere å bruke map, filter eller reduce? (Vet ikke, bare spør)
(map, filter og reduce gjør "hasOwnProperty" automagisk).